std/traits.d

```diff
@@ -2658,6 +2659,150 @@ template RepresentationTypeTuple(T)
+        private template Sort(T...)
+        {
+            template Merge(T...)
+            {
+                template With(U...)
+                {
+                    static if (T.length == 0)
+                        alias With = U;
+                    else static if (U.length == 0)
+                        alias With = T;
+                    else static if (T[0] <= U[0])
+                        alias With =
+                            AliasSeq!(T[0], T[1], Merge!(T[2 .. $]).With!U);
+                    else
+                        alias With =
+                            AliasSeq!(U[0], U[1], Merge!(T).With!(U[2 .. $]));
+                }
+            }
+            static if (T.length <= 2)
+                alias Sort = T;
+            else
+                alias Sort = Merge!(Sort!(T[0 .. T.length / 2]))
+                    .With!(Sort!(T[T.length / 2 .. $]));
+        }
+
+        private alias names = FieldNameTuple!T;
+
+        // Implementation tracks the current index to process in startAt
+        private template Impl(size_t startAt)
+        {
+            static if (names.length <= startAt)
+            {
+                alias Impl = AliasSeq!(); // all done
+            }
+            else
+            {
+                enum size_t offset = mixin("T." ~ names[startAt] ~ ".offsetof");
+                mixin("alias F = typeof(T." ~ names[startAt] ~ ");");
+                alias Impl = AliasSeq!(Add!(offset, Layout!F), Impl!(startAt + 1));
+            }
+        }
+
+        static if (is(T == class))
+            alias Layout = AliasSeq!(0, __Vtable!T*, size_t.sizeof, __Bookkeeping!T*, Sort!(Impl!0));
+        else static if (is(T == interface))
+            alias Layout = AliasSeq!(0, __Vtable!T*, Sort!(Impl!0));
+        else
+            alias Layout = AliasSeq!(Sort!(Impl!0));
+    }
+}
+
+///
+@safe unittest
+{
+    struct S1 { int a; float b; }
+    alias L1 = Layout!S1;
+    static assert(L1.length == 4);
+    static assert(L1[0] == 0 && is(L1[1] == int));
+    static assert(L1[2] == 4 && is(L1[3] == float));
+
+    struct S2 { char[] a; union { S1 b; S1 * c; } }
+    alias L2 = Layout!S2;
+    static assert(L2.length == 8);
+    static assert(L2[0] == 0 && is(L2[1] == char[]));
+    static assert(L2[2] == 16 && is(L2[3] == int));
+    static assert(L2[4] == 16 && is(L2[5] == S1*));
+    static assert(L2[6] == 20 && is(L2[7] == float));
+
+    class C1 { char[] a; union { S1 b; S1 * c; } }
+    alias L3 = Layout!C1;
+    static assert(L3.length == 12);
+    static assert(L3[0] == 0 && is(L3[1] == __Vtable!C1*));
+    static assert(L3[2] == size_t.sizeof && is(L3[3] == __Bookkeeping!C1*));
+    static assert(L3[4] == 2 * size_t.sizeof && is(L3[5] == char[]));
+    static assert(L3[6] == 4 * size_t.sizeof && is(L3[7] == int));
+    static assert(L3[8] == 4 * size_t.sizeof && is(L3[9] == S1*));
+    static assert(L3[10] == 4 + 4 * size_t.sizeof && is(L3[11] == float));
```

Comment:
Ideas for tests missing:
- `align(1) struct S { byte a; int b; byte c; int d;}`
- `align(2) struct S { byte a; int b; byte c; int d;}`
- `struct S { byte a; align(1) int b; byte c; int d;}`
- Test `Layout!(const S2)`. Specify that Layout discards all qualifiers on the types, also of struct members. Test with `struct S1 { int a; const S2 b}` (note the inner const)
- Test with abstract class, class inheritance, interfaces.
- extern(C++) types
- `Layout!int` and `Layout!(int[])` and `Layout!(int[5])` (test errors or sane behavior)
- Sorting for union of these structs: `align(1) struct S1 { byte a; int b; char c; char d; }; align(1) struct S2 { short a; short b; long c; byte d; }; align(1) struct S3 { int a; double b; };`   (intentionally many different types, but avoiding pointers because `size_t` is variable).
- Perhaps important to take into account when testing that `size_t` can be 16bits.